### PR TITLE
Store booking bug data

### PR DESCRIPTION
**Pull request by dwhenry**
Data is stored in three ways:
* Current view of values (Transaction)
* Historical versions of values (TransactionVersion)
* Monthly Historical snapshots of the data (TransactionSummary)

### CODE REVIEW

File: app/models/transaction.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  validates :uid, :booked_at, :booking_at, :booking_status, :delivery_partner,
+            presence: true
+  validates :uid, uniqueness: true
+
+  scope :transactions, ->(end_point) { where(['booking_at <= ?', end_point]) }
+  scope :bookings, ->(end_point) { where(['booked_at <= ?', end_point]).where(cancelled: false) }
+  scope :completed_transactions, ->(end_point) { transactions(end_point).where(booking_status: 'Complete') }
+  scope :new_today, -> { where(updated_at: Time.zone.today..Time.zone.now, version: 1) }
+
+  def set_version
```
Comment (benlovell) [inline app/models/transaction.rb]: May make sense to default the `version` to `0` and always increment on `before_save`?

**Comment by dwhenry** [comment]
@benlovell feel this is now ready to attempt the "Ben Lovell ™️ pedantry check!" 😄

File: app/models/transaction.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+class Transaction < ActiveRecord::Base
+  before_save :increment_version, :create_version, if: :changed?
+
+  validates :uid, :booked_at, :booking_at, :booking_status, :delivery_partner,
+            presence: true
+  validates :uid, uniqueness: true
+  validates :delivery_partner, inclusion: { in: Partners.delivery_partners }
+
+  scope :bookings, ->(end_point) { where(['booked_at <= ?', end_point]).where(cancelled: false) }
```
Comment (benlovell) [inline app/models/transaction.rb]: Pretty sure you don't need to wrap the parameters in an array here, and in line 11.

File: db/migrate/20160610105323_create_transaction_versions.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+  def change
+    create_table :transaction_versions do |t|
+      t.string :uid, default: '', null: false
+      t.datetime :booked_at, null: false
+      t.datetime :booking_at, null: false
+      t.boolean :cancelled, default: false
+      t.string :booking_status, default: '', null: false
+      t.string :delivery_partner, default: '', null: false
+      t.integer :version, default: 0, null: false
+
```
Comment (benlovell) [inline db/migrate/20160610105323_create_transaction_versions.rb]: Extra whitespace line here.

File: app/models/transaction.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+class Transaction < ActiveRecord::Base
```
Comment (benlovell) [inline app/models/transaction.rb]: This is probably not a good name. Also, you may have redefined AR's `Transaction` here...

File: spec/factories.rb [hunk 5]
```diff
@@ -1,4 +1,16 @@
 FactoryGirl.define do
+  factory :transaction_summary do
```
Comment (benlovell) [inline spec/factories.rb]: Could we split each factory with a newline just to separate things?

File: app/models/appointment.rb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+class Appointment < ActiveRecord::Base
+  before_save :increment_version, :create_version, if: :changed?
+
+  validates :uid, :booked_at, :booking_at, :booking_status, :delivery_partner,
+            presence: true
+  validates :uid, uniqueness: true
+  validates :delivery_partner, inclusion: { in: Partners.delivery_partners }
+
+  scope :bookings, ->(end_point) { where('booked_at <= ?', end_point).where(cancelled: false) }
+  scope :completed_transactions, ->(end_point) { transactions(end_point).where(booking_status: 'Complete') }
```
Comment (benlovell) [inline app/models/appointment.rb]: Are we going to keep the `transaction` language around now we've renamed elsewhere?

**Comment by benlovell** [comment]
🚢

**Pull request merged**